From 69cc6ab11a8958670014369d4afa8923530cf3a4 Mon Sep 17 00:00:00 2001
From: Naresh Solanki <naresh.solanki@9elements.com>
Date: Thu, 19 Mar 2026 13:26:30 +0000
Subject: [PATCH 1/9] schemas: legacy: Add MuxChannel object

Mux devices are allowed to be bound to driver when configured in
Entity-Manager. It now creates i2c buses for the mux channels, and
links the bus path with /dev/i2c-mux/<MuxName>/<ChannelName> for other
applications to use without having to know which logical bus maps to
which channel. However, there's currently no way for sensors configured
in EM to use the mux channel buses created in their "Bus" config if
they don't have an IPMI-formatted FRU to be scanned.

This patch proposes an object `MuxChannel` with 2 properties `MuxName`
and `ChannelName` that an i2c sensor in `Exposes` can use to configure
which channel name of which mux they are present at.

Upstream-URL: https://gerrit.openbmc.org/c/openbmc/entity-manager/+/69030
Upstream-Status: Submitted
Signed-off-by: Chau Ly <chaul@amperecomputing.com>
Change-Id: I3cf0eead9f8d8246b182f8681de5fcd4427a5325


---
 schemas/cpld.json     |  13 +++-
 schemas/firmware.json |  25 +++++++-
 schemas/legacy.json   | 145 +++++++++++++++++++++++++++++++++++++++---
 schemas/mctp.json     |  26 +++++++-
 4 files changed, 194 insertions(+), 15 deletions(-)

diff --git a/schemas/cpld.json b/schemas/cpld.json
index f33c22b07..e649f87fd 100644
--- a/schemas/cpld.json
+++ b/schemas/cpld.json
@@ -12,6 +12,9 @@
                 "Bus": {
                     "$ref": "legacy.json#/$defs/Types/Bus"
                 },
+                "MuxChannel": {
+                    "$ref": "legacy.json#/$defs/Types/MuxChannel"
+                },
                 "FirmwareInfo": {
                     "$ref": "firmware.json#/$defs/FirmwareInfoDef"
                 },
@@ -28,7 +31,15 @@
                     ]
                 }
             },
-            "required": ["Address", "Bus", "FirmwareInfo", "Name", "Type"]
+            "required": ["Address", "FirmwareInfo", "Name", "Type"],
+            "oneOf": [
+                {
+                    "required": ["Bus"]
+                },
+                {
+                    "required": ["MuxChannel"]
+                }
+            ]
         }
     }
 }
diff --git a/schemas/firmware.json b/schemas/firmware.json
index 9b559e9d9..a1d7a9c5c 100644
--- a/schemas/firmware.json
+++ b/schemas/firmware.json
@@ -32,6 +32,9 @@
                     "description": "I2C bus, the device is connected to",
                     "type": "integer"
                 },
+                "MuxChannel": {
+                    "$ref": "legacy.json#/$defs/Types/MuxChannel"
+                },
                 "Address": {
                     "description": "I2C address of the device",
                     "type": "integer"
@@ -40,7 +43,15 @@
                     "$ref": "#/$defs/FirmwareInfoDef"
                 }
             },
-            "required": ["Name", "Type", "Bus", "Address", "FirmwareInfo"]
+            "required": ["Name", "Type", "Address", "FirmwareInfo"],
+            "oneOf": [
+                {
+                    "required": ["Bus"]
+                },
+                {
+                    "required": ["MuxChannel"]
+                }
+            ]
         },
         "BIOS": {
             "description": "Required information to complete the host fw update process.",
@@ -107,6 +118,9 @@
                 "Bus": {
                     "$ref": "legacy.json#/$defs/Types/Bus"
                 },
+                "MuxChannel": {
+                    "$ref": "legacy.json#/$defs/Types/MuxChannel"
+                },
                 "Address": {
                     "$ref": "legacy.json#/$defs/Types/Address"
                 },
@@ -137,10 +151,17 @@
             "required": [
                 "Name",
                 "Type",
-                "Bus",
                 "Address",
                 "FirmwareDevice",
                 "FirmwareInfo"
+            ],
+            "oneOf": [
+                {
+                    "required": ["Bus"]
+                },
+                {
+                    "required": ["MuxChannel"]
+                }
             ]
         }
     }
diff --git a/schemas/legacy.json b/schemas/legacy.json
index e54822839..d255869c8 100644
--- a/schemas/legacy.json
+++ b/schemas/legacy.json
@@ -122,6 +122,9 @@
                 "Bus": {
                     "$ref": "#/$defs/Types/Bus"
                 },
+                "MuxChannel": {
+                    "$ref": "#/$defs/Types/MuxChannel"
+                },
                 "Class": {
                     "$ref": "#/$defs/Types/Class"
                 },
@@ -140,6 +143,7 @@
                 }
             },
             "required": ["Name", "Type", "Class"],
+            "not": { "required": ["Bus", "MuxChannel"] },
             "type": "object"
         },
         "AspeedFan": {
@@ -195,6 +199,9 @@
                 "Bus": {
                     "$ref": "#/$defs/Types/Bus"
                 },
+                "MuxChannel": {
+                    "$ref": "#/$defs/Types/MuxChannel"
+                },
                 "EntityId": {
                     "$ref": "#/$defs/Types/EntityId"
                 },
@@ -228,6 +235,7 @@
                 }
             },
             "required": ["Name", "Type"],
+            "not": { "required": ["Bus", "MuxChannel"] },
             "type": "object"
         },
         "NuvotonFan": {
@@ -278,6 +286,9 @@
                 "Bus": {
                     "$ref": "#/$defs/Types/Bus"
                 },
+                "MuxChannel": {
+                    "$ref": "#/$defs/Types/MuxChannel"
+                },
                 "PowerState": {
                     "$ref": "#/$defs/Types/PowerState"
                 },
@@ -299,6 +310,7 @@
                 }
             },
             "required": ["Name", "Type", "CpuID", "Address"],
+            "not": { "required": ["Bus", "MuxChannel"] },
             "type": "object"
         },
         "IpmbSensor": {
@@ -348,6 +360,9 @@
                 "Bus": {
                     "$ref": "#/$defs/Types/Bus"
                 },
+                "MuxChannel": {
+                    "$ref": "#/$defs/Types/MuxChannel"
+                },
                 "Address": {
                     "$ref": "#/$defs/Types/Address"
                 },
@@ -364,7 +379,15 @@
                     ]
                 }
             },
-            "required": ["Name", "Type", "Address", "Bus", "ChannelNames"],
+            "required": ["Name", "Type", "Address", "ChannelNames"],
+            "oneOf": [
+                {
+                    "required": ["Bus"]
+                },
+                {
+                    "required": ["MuxChannel"]
+                }
+            ],
             "type": "object"
         },
         "Gpio": {
@@ -416,6 +439,9 @@
                 "Bus": {
                     "$ref": "#/$defs/Types/Bus"
                 },
+                "MuxChannel": {
+                    "$ref": "#/$defs/Types/MuxChannel"
+                },
                 "Address": {
                     "$ref": "#/$defs/Types/Address"
                 },
@@ -424,7 +450,15 @@
                     "const": "MultiNodePresence"
                 }
             },
-            "required": ["Name", "Type", "Bus", "Address"],
+            "required": ["Name", "Type", "Address"],
+            "oneOf": [
+                {
+                    "required": ["Bus"]
+                },
+                {
+                    "required": ["MuxChannel"]
+                }
+            ],
             "type": "object"
         },
         "HostCpuUtilization": {
@@ -524,6 +558,9 @@
                 "Bus": {
                     "$ref": "#/$defs/Types/Bus"
                 },
+                "MuxChannel": {
+                    "$ref": "#/$defs/Types/MuxChannel"
+                },
                 "Address": {
                     "oneOf": [
                         {
@@ -538,7 +575,15 @@
                     "const": "PSUPresence"
                 }
             },
-            "required": ["Name", "Type", "Bus", "Address"],
+            "required": ["Name", "Type", "Address"],
+            "oneOf": [
+                {
+                    "required": ["Bus"]
+                },
+                {
+                    "required": ["MuxChannel"]
+                }
+            ],
             "type": "object"
         },
         "EmmcDevice": {
@@ -563,6 +608,9 @@
                 "Bus": {
                     "$ref": "#/$defs/Types/Bus"
                 },
+                "MuxChannel": {
+                    "$ref": "#/$defs/Types/MuxChannel"
+                },
                 "Address": {
                     "$ref": "#/$defs/Types/Address"
                 },
@@ -576,7 +624,15 @@
                     "const": "IpmbDevice"
                 }
             },
-            "required": ["Name", "Type", "Bus", "Address", "Class"],
+            "required": ["Name", "Type", "Address", "Class"],
+            "oneOf": [
+                {
+                    "required": ["Bus"]
+                },
+                {
+                    "required": ["MuxChannel"]
+                }
+            ],
             "type": "object"
         },
         "IpmbPowerMonitor": {
@@ -588,6 +644,9 @@
                 "Bus": {
                     "$ref": "#/$defs/Types/Bus"
                 },
+                "MuxChannel": {
+                    "$ref": "#/$defs/Types/MuxChannel"
+                },
                 "Address": {
                     "$ref": "#/$defs/Types/Address"
                 },
@@ -598,7 +657,15 @@
                     "const": "IpmbPowerMonitor"
                 }
             },
-            "required": ["Name", "Type", "Bus", "Address", "Class"],
+            "required": ["Name", "Type", "Address", "Class"],
+            "oneOf": [
+                {
+                    "required": ["Bus"]
+                },
+                {
+                    "required": ["MuxChannel"]
+                }
+            ],
             "type": "object"
         },
         "ADC": {
@@ -669,6 +736,9 @@
                 "Bus": {
                     "$ref": "#/$defs/Types/Bus"
                 },
+                "MuxChannel": {
+                    "$ref": "#/$defs/Types/MuxChannel"
+                },
                 "EntityId": {
                     "$ref": "#/$defs/Types/EntityId"
                 },
@@ -735,7 +805,15 @@
                     ]
                 }
             },
-            "required": ["Name", "Type", "Bus", "Address"],
+            "required": ["Name", "Type", "Address"],
+            "oneOf": [
+                {
+                    "required": ["Bus"]
+                },
+                {
+                    "required": ["MuxChannel"]
+                }
+            ],
             "type": "object"
         },
         "NVME1000": {
@@ -748,6 +826,9 @@
                 "Bus": {
                     "$ref": "#/$defs/Types/Bus"
                 },
+                "MuxChannel": {
+                    "$ref": "#/$defs/Types/MuxChannel"
+                },
                 "Address": {
                     "$ref": "#/$defs/Types/Address"
                 },
@@ -761,7 +842,15 @@
                     "const": "NVME1000"
                 }
             },
-            "required": ["Name", "Type", "Bus", "Address"],
+            "required": ["Name", "Type", "Address"],
+            "oneOf": [
+                {
+                    "required": ["Bus"]
+                },
+                {
+                    "required": ["MuxChannel"]
+                }
+            ],
             "type": "object"
         },
         "EEPROM": {
@@ -773,6 +862,9 @@
                 "Bus": {
                     "$ref": "#/$defs/Types/Bus"
                 },
+                "MuxChannel": {
+                    "$ref": "#/$defs/Types/MuxChannel"
+                },
                 "Address": {
                     "$ref": "#/$defs/Types/Address"
                 },
@@ -791,7 +883,15 @@
                     ]
                 }
             },
-            "required": ["Name", "Type", "Bus", "Address"],
+            "required": ["Name", "Type", "Address"],
+            "oneOf": [
+                {
+                    "required": ["Bus"]
+                },
+                {
+                    "required": ["MuxChannel"]
+                }
+            ],
             "type": "object"
         },
         "PSUSensor": {
@@ -860,6 +960,9 @@
                 "Bus": {
                     "$ref": "#/$defs/Types/Bus"
                 },
+                "MuxChannel": {
+                    "$ref": "#/$defs/Types/MuxChannel"
+                },
                 "Address": {
                     "$ref": "#/$defs/Types/Address"
                 },
@@ -980,6 +1083,7 @@
                 }
             },
             "required": ["Name", "Type"],
+            "not": { "required": ["Bus", "MuxChannel"] },
             "type": "object"
         },
         "NMSensor": {
@@ -1031,10 +1135,14 @@
                 "Bus": {
                     "$ref": "#/$defs/Types/Bus"
                 },
+                "MuxChannel": {
+                    "$ref": "#/$defs/Types/MuxChannel"
+                },
                 "Type": {
                     "const": "Intel HSBP CPLD"
                 }
-            }
+            },
+            "not": { "required": ["Bus", "MuxChannel"] }
         },
         "GenericSMBusMux": {
             "description": "unknown purpose",
@@ -1075,10 +1183,14 @@
                 "Bus": {
                     "$ref": "#/$defs/Types/Bus"
                 },
+                "MuxChannel": {
+                    "$ref": "#/$defs/Types/MuxChannel"
+                },
                 "Type": {
                     "enum": ["UCD90160", "UCD90320"]
                 }
-            }
+            },
+            "not": { "required": ["Bus", "MuxChannel"] }
         },
         "Types": {
             "Address": {
@@ -1244,6 +1356,19 @@
             "MinValue": {
                 "type": "number"
             },
+            "MuxChannel": {
+                "additionalProperties": false,
+                "properties": {
+                    "MuxName": {
+                        "type": "string"
+                    },
+                    "ChannelName": {
+                        "type": "string"
+                    }
+                },
+                "required": ["MuxName", "ChannelName"],
+                "type": "object"
+            },
             "Name": {
                 "type": "string"
             },
diff --git a/schemas/mctp.json b/schemas/mctp.json
index 7dc16292b..6d8b9ccce 100644
--- a/schemas/mctp.json
+++ b/schemas/mctp.json
@@ -11,6 +11,9 @@
                     "description": "The bus index to which the device is attached",
                     "type": "string"
                 },
+                "MuxChannel": {
+                    "$ref": "legacy.json#/$defs/Types/MuxChannel"
+                },
                 "Name": {
                     "description": "A recognisable name for the device",
                     "type": "string"
@@ -20,7 +23,15 @@
                     "enum": ["MCTPI2CTarget"]
                 }
             },
-            "required": ["Address", "Bus", "Name", "Type"],
+            "required": ["Address", "Name", "Type"],
+            "oneOf": [
+                {
+                    "required": ["Bus"]
+                },
+                {
+                    "required": ["MuxChannel"]
+                }
+            ],
             "type": "object"
         },
         "MCTPI3CTarget": {
@@ -33,6 +44,9 @@
                     "description": "The bus index to which the device is attached",
                     "type": "string"
                 },
+                "MuxChannel": {
+                    "$ref": "legacy.json#/$defs/Types/MuxChannel"
+                },
                 "Name": {
                     "description": "A recognisable name for the device",
                     "type": "string"
@@ -42,7 +56,15 @@
                     "enum": ["MCTPI3CTarget"]
                 }
             },
-            "required": ["Address", "Bus", "Name", "Type"],
+            "required": ["Address", "Name", "Type"],
+            "oneOf": [
+                {
+                    "required": ["Bus"]
+                },
+                {
+                    "required": ["MuxChannel"]
+                }
+            ],
             "type": "object"
         }
     }
-- 
2.43.0

